layer1/executor/tasks/tests/snapshot_test.go

```diff
@@ -0,0 +1,450 @@
+		assert.Nil(t, err)
+		err = snapshotTask.Prepare(ctx)
+		assert.Nil(t, err)
+
+		snapshotState, err := state.GetSnapshotState(snapshotTask.GetDB())
+		assert.Nil(t, err)
+
+		shouldExecute, err := snapshotTask.ShouldExecute(ctx)
+		if shouldExecute {
+			txn, taskError := snapshotTask.Execute(ctx)
+			amILeading, err := utils.AmILeading(
+				suite.Eth,
+				ctx,
+				fixture.Logger,
+				snapshotState.LastSnapshotHeight,
+				snapshotState.RandomSeedHash,
+				snapshotTask.NumOfValidators,
+				snapshotTask.ValidatorIndex,
+				snapshotState.DesperationFactor,
+				snapshotState.DesperationDelay,
+			)
+			assert.Nil(t, err)
+			if amILeading {
+				assert.Nil(t, taskError)
+				rcptResponse, err := fixture.Watcher.Subscribe(ctx, txn, nil)
+				assert.Nil(t, err)
+				tests.WaitGroupReceipts(t, suite.Eth, []transaction.ReceiptResponse{rcptResponse})
+			} else {
+				assert.Nil(t, txn)
+				require.NotNil(t, taskError)
+				assert.True(t, taskError.IsRecoverable())
+			}
+		}
+	}
+}
+
+func Test_LeaderElectionWithRandomData(t *testing.T) {
+	valMin := 4
+	valMax := 50
+	fixture := setupEthereum(t, 5)
+	eth := fixture.Client
+	ctx := context.Background()
+	for i := 0; i < 100; i++ {
+		valNum := getRandomNumberInRange(valMin, valMax)
+		valIndex := getRandomNumberInRange(0, valNum)
+		hash := []byte(uuid.New().String())
+		keccakedHash := ethCrypto.Keccak256(hash)
+		var hashSlice32 [32]byte
+		copy(hashSlice32[:], keccakedHash)
+
+		currentHeight, err := eth.GetCurrentHeight(ctx)
+		assert.Nil(t, err)
+
+		blocksToMine := getRandomNumberInRange(1, 1000)
+		advanceToBlock := currentHeight + uint64(blocksToMine)
+		tests.AdvanceTo(eth, advanceToBlock)
+
+		accounts := eth.GetKnownAccounts()
+		owner := accounts[0]
+		callOpts, err := eth.GetCallOpts(ctx, owner)
+		assert.Nil(t, err)
+
+		start := getRandomNumberInRange(int(advanceToBlock)-128, int(advanceToBlock))
+		desperationFactor, err := fixture.Contracts.EthereumContracts().Snapshots().GetSnapshotDesperationFactor(callOpts)
+		assert.Nil(t, err)
+
+		desperationDelay, err := fixture.Contracts.EthereumContracts().Snapshots().GetSnapshotDesperationDelay(callOpts)
+		assert.Nil(t, err)
+
+		blocksSinceDesperation := utils.GetBlocksSinceDesperation(int(advanceToBlock), start, int(desperationDelay.Int64()))
+		goResult, err := utils.AmILeading(eth, ctx, fixture.Logger, start, keccakedHash, valNum, valIndex, int(desperationFactor.Int64()), int(desperationDelay.Int64()))
+		assert.Nil(t, err)
+
+		solidityResult, err := fixture.Contracts.EthereumContracts().Snapshots().MayValidatorSnapshot(callOpts, big.NewInt(int64(valNum)), big.NewInt(int64(valIndex)), big.NewInt(int64(blocksSinceDesperation)), hashSlice32, desperationFactor)
+		assert.Nil(t, err)
+
+		assert.Equal(t, solidityResult, goResult)
+	}
+}
+
+func Test_LeaderElectionWithFixedData(t *testing.T) {
```

Comment:
Consider marking tests as `t.Parallel()` to speed up execution.

Reply:
added for fixed and random data tests